timeline_shifts_visualization_experiment/02_parametric_vs_aligned_umap.py

```diff
@@ -0,0 +1,389 @@
+            ),
+        )
+    )
+
+relations = []
+for year_one, year_two in tqdm.tqdm(window(training_word_model_map, 2)):
+    year_one_map = training_word_model_map[year_one]["index_to_token_map"]
+    year_two_map = training_word_model_map[year_two]["index_to_token_map"]
+    relations.append(
+        {
+            year_one_map[tok]: year_two_map[tok]
+            for tok in year_one_map
+            if tok in year_two_map
+        }
+    )
+
+words_to_visualize = []
+token_character_list = []
+year_labels_list = []
+
+for year in tqdm.tqdm(training_word_model_map):
+    model = training_word_model_map[year]["model"]
+    word_subset_matrix = model[
+        model.index_to_key[: training_word_model_map[year]["cutoff_index"]]
+    ]
+    print((year, word_subset_matrix.shape))
+    words_to_visualize.append(word_subset_matrix)
+    token_character_list += list(
+        map(
+            lambda x: re.escape(x),
+            model.index_to_key[: training_word_model_map[year]["cutoff_index"]],
+        )
+    )
+    year_labels_list += [year] * len(
+        model.index_to_key[: training_word_model_map[year]["cutoff_index"]]
+    )
+
+aligned_umap_output_path = Path("output/aligned_umap_model.pkl")
+if not aligned_umap_output_path.exists():
+    # runtime - 3h 44m 17s
+    aligned_umap_model = AlignedUMAP(
+        n_components=2,
+        verbose=True,
+        metric="cosine",
+        random_state=101,
+        low_memory=True,
+        n_neighbors=25,
+        min_dist=0.99,
+        alignment_window_size=2,
+        alignment_regularisation=1e-3,
+    )
+    aligned_umap_model.fit(words_to_visualize, relations=relations)
+    aligned_umap_model.embeddings_ = list(aligned_umap_model.embeddings_)
+    pickle.dump(aligned_umap_model, open(str(aligned_umap_output_path), "wb"))
+else:
+    aligned_umap_model = pickle.load(open(str(aligned_umap_output_path), "rb"))
+
+aligned_plot_df = pd.DataFrame(
+    np.vstack(aligned_umap_model.embeddings_), columns=["umap1", "umap2"]
+) >> ply.define(year="year_labels_list", tok="token_character_list")
+aligned_plot_df >> ply.slice_rows(10)
+
+g = (
+    p9.ggplot(
+        aligned_plot_df >> ply.define(year="pd.Categorical(year)"),
+        p9.aes(x="umap1", y="umap2", fill="year"),
+    )
+    + p9.geom_point(alpha=0.05)
+    + p9.labs(title="Aligned UMAP 2010-2020")
+    + p9.theme(figure_size=(10, 8))
+)
+print(g)
+
+# # Take home messages
+
+# 1. Parametric UMAP works as expected. It creates clusters based on training data and uses this training data to map new data onto those clusters. Observed this when training on odd years but projecting even years onto that.
+# 2. Align umap might be better in terms of projections as it constructs individual umap models then aligns them overall.
+# 3. Not sure which model to choose but I believe that parametric umap might be the better but will see as this project unfolds.
```

Comment:
I'm a bit confused on what the end goal is, is it to have a visualization or to use the embedding? And is the idea that if you use parametric UMAP you would train it on a subset of all years then project the held-out set onto that? Or would it be to have "anchor" years that you project onto (like what you did in the notebook where you projected even years onto 2015, 2017, 2019)?

Reply:
>I'm a bit confused on what the end goal is, is it to have a visualization or to use the embedding?

The original goal here is to have a visualization showing how a token shifts through time. Most of these experiments here is to see how UMAP projects tokens into a low dimensional space. I can't use regular UMAP as I run into memory issues (can't fit all ten models that are >10000k words by 300 dimensions) into a 64GB machine. Here I was trying to see if parametric UMAP or align UMAP could be better alternatives to help circumvent this bottleneck. 

>And is the idea that if you use parametric UMAP you would train it on a subset of all years then project the held-out set onto that? Or would it be to have "anchor" years that you project onto (like what you did in the notebook where you projected even years onto 2015, 2017, 2019)?

Yeah for parametric UMAP the idea was to see if I can use odd years as an anchor and then project even years onto the same space. If this was possible, then i could accomplish the above goal. However, the results suggest that align UMAP (create independent umap models the use an alignment algorithm such as procrustes to align at the end) is a better idea than parametric.

Reply:
aha, ok I see 👍🏼